File: flowshop/flowshop.cpp
/********** flowshop.cpp **********/

#include <iostream>
#include <fstream>
#include <vector>
#include "localsolver.h"

using namespace localsolver;
using namespace std;

class Flowshop {
public:
    // Number of jobs
    int nbJobs;

    // Number of machines
    int nbMachines;

    // Initial seed used to generate the instance
    long initialSeed;

    // Upper bound
    int upperBound;

    // Lower bound
    int lowerBound;

    // Processing time
    vector<vector<lsint> > processingTime;

    // LocalSolver
    LocalSolver localsolver;

    // Decision variable
    LSExpression jobs;

    // Objective
    LSExpression makespan;

    // Reads instance data.
    void readInstance(const string& fileName) {
        ifstream infile;
        infile.exceptions(ifstream::failbit | ifstream::badbit);
        infile.open(fileName.c_str());

        infile >> nbJobs;
        infile >> nbMachines;
        infile >> initialSeed;
        infile >> upperBound;
        infile >> lowerBound;

        processingTime.resize(nbMachines);
        for (int m = 0; m < nbMachines; m++) {
            processingTime[m].resize(nbJobs);
            for (int j = 0; j < nbJobs; j++) {
                infile >> processingTime[m][j];
            }
        }
    }

    void solve(int limit) {
        // Declares the optimization model.
        LSModel model = localsolver.getModel();

        // Permutation of jobs
        jobs = model.listVar(nbJobs);

        // All jobs have to be assigned
        model.constraint(model.count(jobs) == nbJobs);

        // For each machine create proccessingTime[m] as an array to be able to access it 
        // with an 'at' operator
        vector<LSExpression> processingTimeArray(nbMachines);
        for (int m = 0; m < nbMachines; m++) {
            processingTimeArray[m] = model.array(processingTime[m].begin(), processingTime[m].end());
        }

        // On machine 0, the jth job ends on the time it took to be processed after
        // the end of the previous job
        vector<LSExpression> end(nbMachines);
        LSExpression firstEndSelector = model.createLambdaFunction([&](LSExpression i, LSExpression prev) {
            return prev + processingTimeArray[0][jobs[i]];
        });
        end[0] = model.array(model.range(0, nbJobs), firstEndSelector);

        // The jth job on machine m starts when it has been processed by machine n-1
        // AND when job j-1 has been processed on machine m. It ends after it has been processed.
        for (int m = 1; m < nbMachines; ++m) {
            int mL = m;
            LSExpression endSelector = model.createLambdaFunction([&](LSExpression i, LSExpression prev) {
                return model.max(prev, end[mL - 1][i]) + processingTimeArray[mL][jobs[i]];
            });
            end[m] = model.array(model.range(0, nbJobs), endSelector);
        }

        // Minimize the makespan: end of the last job on the last machine
        makespan = end[nbMachines - 1][nbJobs - 1];
        model.minimize(makespan);
        model.close();


        // Parameterizes the solver.
        localsolver.getParam().setTimeLimit(limit);

        localsolver.solve();
    }

    // Writes the solution in a file
    void writeSolution(const string& fileName) {
        ofstream outfile;
        outfile.exceptions(ofstream::failbit | ofstream::badbit);
        outfile.open(fileName.c_str());

        outfile << makespan.getValue() << endl;
        LSCollection jobsCollection = jobs.getCollectionValue();
        for (int j = 0; j < nbJobs; j++) {
            outfile << jobsCollection[j] << " ";
        }
        outfile << endl;
    }
};

int main(int argc, char** argv) {
    if (argc < 2) {
        cerr << "Usage: flowshop inputFile [outputFile] [timeLimit]" << endl;
        return 1;
    }

    const char* instanceFile = argv[1];
    const char* solFile = argc > 2 ? argv[2] : NULL;
    const char* strTimeLimit = argc > 3 ? argv[3] : "5";

    try {
        Flowshop model;
        model.readInstance(instanceFile);
        model.solve(atoi(strTimeLimit));
        if (solFile != NULL) model.writeSolution(solFile);
        return 0;
     } catch (const exception& e) {
        cerr << "An error occurred: " << e.what() << endl;
        return 1;
    }
}

File: flowshop/flowshop.py
########## flowshop.py ##########

import localsolver
import sys

if len(sys.argv) < 2:
    print("Usage: python flowshop.py inputFile [outputFile] [timeLimit]")
    sys.exit(1)


def read_integers(filename):
    with open(filename) as f:
        return [int(elem) for elem in f.read().split()]


with localsolver.LocalSolver() as ls:

    #
    # Reads instance data
    #
    file_it = iter(read_integers(sys.argv[1]))

    nb_jobs = int(next(file_it))
    nb_machines = int(next(file_it))
    initial_seed = int(next(file_it))
    upper_bound = int(next(file_it))
    lower_bound = int(next(file_it))

    processing_time = [[int(next(file_it)) for j in range(nb_jobs)] for j in range(nb_machines)]

    #
    # Declares the optimization model
    #
    model = ls.model

    # Permutation of jobs
    jobs = model.list(nb_jobs)

    # All jobs have to be assigned
    model.constraint(model.eq(model.count(jobs), nb_jobs))

    # For each machine create proccessingTime[m] as an array to be able to access it
    # with an 'at' operator
    processing_time_array = [model.array(processing_time[m]) for m in range(nb_machines)]

    # On machine 0, the jth job ends on the time it took to be processed after the end of the previous job
    end = [None]*nb_machines

    first_end_selector = model.lambda_function(lambda i, prev: prev + processing_time_array[0][jobs[i]])

    end[0] = model.array(model.range(0, nb_jobs), first_end_selector)

    # The jth job on machine m starts when it has been processed by machine n-1
    # AND when job j-1 has been processed on machine m. It ends after it has been processed.
    for m in range(1, nb_machines):
        mL = m
        end_selector = model.lambda_function(lambda i, prev: \
                    model.max(prev, end[mL - 1][i]) + \
                    processing_time_array[mL][jobs[i]])
        end[m] = model.array(model.range(0, nb_jobs), end_selector)


    # Minimize the makespan: end of the last job on the last machine
    makespan = end[nb_machines-1][nb_jobs-1]
    model.minimize(makespan)

    model.close()

    #
    # Parameterizes the solver
    #
    if len(sys.argv) >= 4: ls.param.time_limit = int(sys.argv[3])
    else: ls.param.time_limit = 5

    ls.solve()

    #
    # Writes the solution in a file
    #
    if len(sys.argv) >= 3:
        with open(sys.argv[2], 'w') as f:
            f.write("%d\n" % makespan.value)
            for j in jobs.value:
                f.write("%d " % j)
            f.write("\n")
